ext/opcache/ZendAccelerator.c

```diff
@@ -3020,6 +3227,615 @@ void accelerator_shm_read_unlock(void)
+
+		if (EXPECTED(ce->type == ZEND_USER_CLASS)) {
+			_zend_hash_append_ptr(dst, p->key, ce);
+			zend_hash_del_bucket(src, p);
+		} else {
+			break;
+		}
+	} ZEND_HASH_FOREACH_END();
+	src->pDestructor = orig_dtor;
+}
+
+static zend_op_array *preload_compile_file(zend_file_handle *file_handle, int type)
+{
+	zend_op_array *op_array = preload_orig_compile_file(file_handle, type);
+
+	if (op_array && op_array->refcount) {
+		zend_persistent_script *script;
+
+		script = create_persistent_script();
+		script->script.first_early_binding_opline = (uint32_t)-1;
+		script->script.filename = zend_string_copy(op_array->filename);
+		zend_string_hash_val(script->script.filename);
+		script->script.main_op_array = *op_array;
+
+//???		efree(op_array->refcount);
+		op_array->refcount = NULL;
+
+		if (op_array->static_variables &&
+		    !(GC_FLAGS(op_array->static_variables) & IS_ARRAY_IMMUTABLE)) {
+			GC_ADDREF(op_array->static_variables);
+		}
+
+		zend_hash_add_ptr(preload_scripts, script->script.filename, script);
+	}
+
+	return op_array;
+}
+
+static void preload_sort_classes(void *base, size_t count, size_t siz, compare_func_t compare, swap_func_t swp) /* {{{ */
+{
+	Bucket *b1 = base;
+	Bucket *b2;
+	Bucket *end = b1 + count;
+	Bucket tmp;
+	zend_class_entry *ce, *p;
+
+	while (b1 < end) {
+try_again:
+		ce = (zend_class_entry*)Z_PTR(b1->val);
+		if (ce->parent && (ce->ce_flags & ZEND_ACC_LINKED)) {
+			p = ce->parent;
+			if (p->type == ZEND_USER_CLASS) {
+				b2 = b1 + 1;
+				while (b2 < end) {
+					if (p ==  Z_PTR(b2->val)) {
+						tmp = *b1;
+						*b1 = *b2;
+						*b2 = tmp;
+						goto try_again;
+					}
+					b2++;
+				}
+			}
+		}
+		if (ce->num_interfaces && (ce->ce_flags & ZEND_ACC_LINKED)) {
+			uint32_t i = 0;
+			for (i = 0; i < ce->num_interfaces; i++) {
+				p = ce->interfaces[i];
+				if (p->type == ZEND_USER_CLASS) {
+					b2 = b1 + 1;
+					while (b2 < end) {
+						if (p ==  Z_PTR(b2->val)) {
+							tmp = *b1;
+							*b1 = *b2;
+							*b2 = tmp;
+							goto try_again;
+						}
+						b2++;
+					}
+				}
+			}
+		}
+		b1++;
+	}
+}
+
+static void preload_link(void)
+{
+	zval *zv;
+	zend_persistent_script *script;
+	zend_class_entry *ce, *parent, *p;
+	zend_string *key;
+	zend_bool found, changed;
+	uint32_t i;
+
+	/* Resolve class dependencies */
+	do {
+		changed = 0;
+
+		ZEND_HASH_REVERSE_FOREACH_VAL(EG(class_table), zv) {
+			ce = Z_PTR_P(zv);
+			if (ce->type == ZEND_INTERNAL_CLASS) {
+				break;
+			}
+			if ((ce->ce_flags & ZEND_ACC_TOP_LEVEL)
+			 && !(ce->ce_flags & ZEND_ACC_LINKED)) {
+
+				parent = NULL;
+
+				if (ce->parent_name) {
+					key = zend_string_tolower(ce->parent_name);
+					parent = zend_hash_find_ptr(EG(class_table), key);
+					zend_string_release(key);
+					if (!parent) continue;
+				}
+
+				if (ce->num_interfaces) {
+					found = 1;
+					for (i = 0; i < ce->num_interfaces; i++) {
+						p = zend_hash_find_ptr(EG(class_table), ce->interface_names[i].lc_name);
```

Comment:
Does `EG(class_table)` also include internal classes at this point? If so, I'm wondering how this will be handled on Windows where IIRC multiple processes (with internal classes at different addrs) may attach to one shm instance.

Reply:
Yes. I know about this problem, but prefer to postpone it. e.g. don't support this feature on Windows or copy internal classes into SHM.